File: net/net.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <termios.h>
#include <sys/select.h>
#include <curl/curl.h>
#include <stdatomic.h>

// Target: 600GB
#define TARGET_GB 600
#define TARGET_BYTES ((unsigned long long)TARGET_GB * 1024ULL * 1024ULL * 1024ULL)
#define CHUNK_SIZE (512 * 1024)  // 512KB chunks (optimal for network)
#define THREADS_PER_URL 4  // More parallel connections per source

// Passive mode settings
#define DOWNLOAD_DURATION 60  // seconds to download actively
#define SLEEP_DURATION 10  // seconds to sleep between bursts

// Test file URLs
static const char *TEST_FILES[] = {
    "http://speedtest.tele2.net/10GB.zip",
    "http://speedtest.tele2.net/1GB.zip",
    "http://proof.ovh.net/files/10Gb.dat",
    "http://proof.ovh.net/files/1Gb.dat",
    "http://speedtest.belwue.net/100M",
    "http://speedtest-sgp1.digitalocean.com/10gb.test",
    "http://speedtest-ams2.digitalocean.com/10gb.test",
    "http://speedtest.ftp.otenet.gr/files/test10Gb.db",
};
#define NUM_URLS (sizeof(TEST_FILES) / sizeof(TEST_FILES[0]))
#define NUM_THREADS (NUM_URLS * THREADS_PER_URL)

// Shared state
static atomic_ullong total_bytes = 0;
static atomic_int stop_flag = 0;
static atomic_int pause_flag = 0;
static atomic_int passive_sleep_flag = 0;  // For passive mode sleep cycles
static pthread_mutex_t print_lock = PTHREAD_MUTEX_INITIALIZER;

// Terminal settings for keyboard input
static struct termios old_termios;
static int termios_saved = 0;

// Thread argument structure
typedef struct {
    const char *url;
    int thread_id;
} thread_arg_t;

// Format bytes to human readable
static void format_bytes(unsigned long long bytes, char *buf, size_t buflen) {
    const char *units[] = {"B", "KB", "MB", "GB", "TB", "PB"};
    int unit_idx = 0;
    double value = (double)bytes;

    while (value >= 1024.0 && unit_idx < 5) {
        value /= 1024.0;
        unit_idx++;
    }
    snprintf(buf, buflen, "%.2f %s", value, units[unit_idx]);
}

// Format speed in MB/s
static void format_speed(double bytes_per_sec, char *buf, size_t buflen) {
    double mbps = bytes_per_sec / (1024.0 * 1024.0);
    snprintf(buf, buflen, "%.1f MB/s", mbps);
}

// CURL write callback - just count bytes and discard data
static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    (void)contents;
    (void)userp;
    size_t real_size = size * nmemb;

    // Wait while paused or in passive sleep
    while ((atomic_load(&pause_flag) || atomic_load(&passive_sleep_flag)) && !atomic_load(&stop_flag)) {
        usleep(100000);  // 100ms
    }

    if (atomic_load(&stop_flag)) {
        return 0;  // Signal curl to stop
    }

    atomic_fetch_add(&total_bytes, real_size);
    return real_size;
}

// Download thread function
static void *download_thread(void *arg) {
    thread_arg_t *targ = (thread_arg_t *)arg;
    CURL *curl;
    CURLcode res;

    curl = curl_easy_init();
    if (!curl) {
        pthread_mutex_lock(&print_lock);
        fprintf(stderr, "Thread %d: Failed to init curl\n", targ->thread_id);
        pthread_mutex_unlock(&print_lock);
        free(targ);
        return NULL;
    }

    while (!atomic_load(&stop_flag)) {
        curl_easy_setopt(curl, CURLOPT_URL, targ->url);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, NULL);
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 0L);  // No timeout
        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 10L);  // Faster connection timeout
        curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
        curl_easy_setopt(curl, CURLOPT_BUFFERSIZE, CHUNK_SIZE);

        // Speed optimizations
        curl_easy_setopt(curl, CURLOPT_TCP_NODELAY, 1L);  // Disable Nagle's algorithm
        curl_easy_setopt(curl, CURLOPT_TCP_KEEPALIVE, 1L);  // Keep connections alive
        curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "");  // Accept any encoding (gzip, etc.)
        curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);  // Force HTTP/1.1
        curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 0L);  // Allow connection reuse
        curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 0L);  // Use cached connections
        curl_easy_setopt(curl, CURLOPT_DNS_CACHE_TIMEOUT, 3600L);  // Cache DNS for 1 hour
        curl_easy_setopt(curl, CURLOPT_IPRESOLVE, CURL_IPRESOLVE_V4);  // IPv4 only (often faster)

        res = curl_easy_perform(curl);

        if (res != CURLE_OK && !atomic_load(&stop_flag)) {
            // Brief pause on error, then retry
            sleep(1);
        }
    }

    curl_easy_cleanup(curl);
    free(targ);
    return NULL;
}

// Keyboard listener thread
static void *keyboard_listener(void *arg) {
    (void)arg;
    fd_set readfds;
    struct timeval tv;
    char c;

    while (!atomic_load(&stop_flag)) {
        FD_ZERO(&readfds);
        FD_SET(STDIN_FILENO, &readfds);
        tv.tv_sec = 0;
        tv.tv_usec = 100000;  // 100ms

        if (select(STDIN_FILENO + 1, &readfds, NULL, NULL, &tv) > 0) {
            if (read(STDIN_FILENO, &c, 1) == 1) {
                if (c == 'p' || c == 'P') {
                    int paused = !atomic_load(&pause_flag);
                    atomic_store(&pause_flag, paused);

                    pthread_mutex_lock(&print_lock);
                    if (paused) {
                        printf("\n⏸️  PAUSED - Press 'p' to resume\n");
                    } else {
                        printf("\n▶️  RESUMED - Press 'p' to pause\n");
                    }
                    fflush(stdout);
                    pthread_mutex_unlock(&print_lock);
                } else if (c == 'q' || c == 'Q') {
                    pthread_mutex_lock(&print_lock);
                    printf("\n🛑 Stopping downloads...\n");
                    fflush(stdout);
                    pthread_mutex_unlock(&print_lock);
                    atomic_store(&stop_flag, 1);
                    break;
                }
            }
        }
    }
    return NULL;
}

// Passive mode manager thread
static void *passive_mode_manager(void *arg) {
    (void)arg;
    int cycle = 0;

    while (!atomic_load(&stop_flag)) {
        // Download phase
        pthread_mutex_lock(&print_lock);
        printf("\n🟢 Passive Mode: Starting download burst (cycle %d)\n", ++cycle);
        fflush(stdout);
        pthread_mutex_unlock(&print_lock);

        atomic_store(&passive_sleep_flag, 0);
        sleep(DOWNLOAD_DURATION);

        if (atomic_load(&stop_flag)) break;

        // Sleep phase
        pthread_mutex_lock(&print_lock);
        printf("\n💤 Passive Mode: Sleeping for %d seconds...\n", SLEEP_DURATION);
        fflush(stdout);
        pthread_mutex_unlock(&print_lock);

        atomic_store(&passive_sleep_flag, 1);
        sleep(SLEEP_DURATION);
    }

    return NULL;
}

// Format ETA to human readable
static void format_eta(double seconds, char *buf, size_t buflen) {
    if (seconds < 0 || seconds > 365 * 24 * 3600) {
        snprintf(buf, buflen, "--:--:--");
        return;
    }
    int hours = (int)(seconds / 3600);
    int mins = (int)((seconds - hours * 3600) / 60);
    int secs = (int)(seconds - hours * 3600 - mins * 60);
    snprintf(buf, buflen, "%02d:%02d:%02d", hours, mins, secs);
}

// Progress bar
static void print_progress(unsigned long long bytes, unsigned long long total,
                          double speed, double elapsed_hours, double eta_seconds, int paused, int sleeping) {
    char bytes_str[32], total_str[32], speed_str[32], eta_str[32];
    int bar_width = 30;
    double progress = (double)bytes / (double)total;
    int filled = (int)(progress * bar_width);

    format_bytes(bytes, bytes_str, sizeof(bytes_str));
    format_bytes(total, total_str, sizeof(total_str));
    format_speed(speed, speed_str, sizeof(speed_str));
    format_eta(eta_seconds, eta_str, sizeof(eta_str));

    printf("\r");
    if (sleeping) {
        printf("💤 SLEEPING: ");
    } else if (paused) {
        printf("⏸️  PAUSED : ");
    } else {
        printf("📥 Progress: ");
    }

    printf("%5.1f%%|", progress * 100);
    for (int i = 0; i < bar_width; i++) {
        if (i < filled) printf("█");
        else printf("░");
    }
    printf("| %s/%s | 🚀 %s | ⏱️ %.2fh | ETA: %s",
           bytes_str, total_str, speed_str, elapsed_hours, eta_str);

    if (sleeping) {
        printf(" (passive mode)");
    } else if (paused) {
        printf(" (paused)");
    }

    printf("    ");  // Clear any trailing chars
    fflush(stdout);
}

// Signal handler for clean exit
static void signal_handler(int sig) {
    (void)sig;
    atomic_store(&stop_flag, 1);
}

// Restore terminal settings
static void restore_terminal(void) {
    if (termios_saved) {
        tcsetattr(STDIN_FILENO, TCSADRAIN, &old_termios);
    }
}

// Setup terminal for raw input
static int setup_terminal(void) {
    struct termios new_termios;

    if (tcgetattr(STDIN_FILENO, &old_termios) < 0) {
        return -1;
    }
    termios_saved = 1;
    atexit(restore_terminal);

    new_termios = old_termios;
    new_termios.c_lflag &= ~(ICANON | ECHO);
    new_termios.c_cc[VMIN] = 0;
    new_termios.c_cc[VTIME] = 0;

    if (tcsetattr(STDIN_FILENO, TCSANOW, &new_termios) < 0) {
        return -1;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    pthread_t threads[NUM_THREADS];
    pthread_t kb_thread;
    pthread_t passive_thread;
    int thread_count = 0;
    time_t start_time, current_time;
    unsigned long long last_bytes = 0;
    double pause_time = 0;
    time_t pause_start = 0;
    int passive_mode = 0;  // Default: active mode (normal)

    // Parse command-line arguments
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--passive") == 0) {
            passive_mode = 1;
        } else if (strcmp(argv[i], "--help") == 0 || strcmp(argv[i], "-h") == 0) {
            printf("Usage: %s [options]\n", argv[0]);
            printf("  --passive    Enable passive mode (default: active mode)\n");
            printf("  --help       Show this help message\n");
            return 0;
        }
    }

    // Initialize curl
    curl_global_init(CURL_GLOBAL_ALL);

    // Setup signal handlers
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    // Setup terminal for raw input
    if (setup_terminal() < 0) {
        printf("ℹ️  Keyboard controls not available in this terminal\n");
    }

    // Print header
    printf("============================================================\n");
    printf("🚀 HIGH-SPEED DATA DOWNLOADER (C Version)\n");
    printf("🎯 Target: %d GB\n", TARGET_GB);
    printf("📡 Using %lu download sources with %d threads each\n",
           (unsigned long)NUM_URLS, THREADS_PER_URL);
    printf("============================================================\n");
    printf("⚠️  Data is downloaded to RAM and discarded (no disk usage)\n");
    printf("============================================================\n");
    if (passive_mode) {
        printf("💤 PASSIVE MODE: Download %ds, Sleep %ds\n", DOWNLOAD_DURATION, SLEEP_DURATION);
    } else {
        printf("📥 ACTIVE MODE: Continuous downloading\n");
    }
    printf("============================================================\n");
    printf("🎮 Controls: Press 'p' to PAUSE/PLAY, 'q' to QUIT\n");
    printf("============================================================\n\n");

    start_time = time(NULL);

    // Start keyboard listener thread
    if (pthread_create(&kb_thread, NULL, keyboard_listener, NULL) != 0) {
        fprintf(stderr, "Failed to create keyboard listener thread\n");
    }

    if (passive_mode) {
        // Start passive mode manager thread
        if (pthread_create(&passive_thread, NULL, passive_mode_manager, NULL) != 0) {
            fprintf(stderr, "Failed to create passive mode manager thread\n");
        }
    }

    // Start download threads
    for (size_t i = 0; i < NUM_URLS; i++) {
        for (int j = 0; j < THREADS_PER_URL; j++) {
            thread_arg_t *arg = malloc(sizeof(thread_arg_t));
            if (!arg) continue;

            arg->url = TEST_FILES[i];
            arg->thread_id = thread_count;

            if (pthread_create(&threads[thread_count], NULL, download_thread, arg) == 0) {
                thread_count++;
            } else {
                free(arg);
            }
        }
    }

    printf("Started %d download threads\n\n", thread_count);

    // Progress monitoring loop
    double avg_speed_smooth = 0;  // Smoothed average speed for ETA
    while (!atomic_load(&stop_flag)) {
        unsigned long long current_bytes = atomic_load(&total_bytes);
        current_time = time(NULL);

        // Check if target reached
        if (current_bytes >= TARGET_BYTES) {
            atomic_store(&stop_flag, 1);
            break;
        }

        // Handle pause time tracking
        if (atomic_load(&pause_flag)) {
            if (pause_start == 0) {
                pause_start = current_time;
            }
        } else {
            if (pause_start != 0) {
                pause_time += difftime(current_time, pause_start);
                pause_start = 0;
            }
        }

        // Calculate elapsed time (excluding pause)
        double elapsed = difftime(current_time, start_time);
        if (pause_start != 0) {
            elapsed -= difftime(current_time, pause_start);
        }
        elapsed -= pause_time;
        double elapsed_hours = elapsed / 3600.0;

        // Calculate speed
        double speed = 0;
        if (elapsed > 0) {
            speed = (double)(current_bytes - last_bytes);  // bytes in last second
        }

        // Calculate smoothed average speed for better ETA
        if (avg_speed_smooth == 0 && speed > 0) {
            avg_speed_smooth = speed;
        } else if (speed > 0) {
            avg_speed_smooth = avg_speed_smooth * 0.9 + speed * 0.1;  // Exponential smoothing
        }

        // Calculate ETA
        double eta_seconds = -1;
        if (avg_speed_smooth > 0) {
            unsigned long long remaining_bytes = TARGET_BYTES - current_bytes;
            eta_seconds = (double)remaining_bytes / avg_speed_smooth;
        }

        // Print progress
        pthread_mutex_lock(&print_lock);
        print_progress(current_bytes, TARGET_BYTES, speed, elapsed_hours, eta_seconds,
                      atomic_load(&pause_flag), atomic_load(&passive_sleep_flag));
        pthread_mutex_unlock(&print_lock);

        last_bytes = current_bytes;

        sleep(1);
    }

    printf("\n\nWaiting for threads to finish...\n");

    // Wait for all threads
    for (int i = 0; i < thread_count; i++) {
        pthread_join(threads[i], NULL);
    }
    pthread_join(kb_thread, NULL);
    if (passive_mode) {
        pthread_join(passive_thread, NULL);
    }

    // Calculate final statistics
    unsigned long long final_bytes = atomic_load(&total_bytes);
    double elapsed = difftime(time(NULL), start_time) - pause_time;
    double hours = elapsed / 3600.0;
    double avg_speed = (elapsed > 0) ? (double)final_bytes / elapsed : 0;

    // Print summary
    char bytes_str[32], speed_str[32];
    format_bytes(final_bytes, bytes_str, sizeof(bytes_str));
    format_speed(avg_speed, speed_str, sizeof(speed_str));

    printf("\n============================================================\n");
    if (final_bytes >= TARGET_BYTES) {
        printf("🎉 TARGET REACHED: %d GB Downloaded!\n", TARGET_GB);
    } else {
        printf("📊 Downloaded: %s\n", bytes_str);
    }
    printf("============================================================\n");
    printf("📥 Total Downloaded: %s\n", bytes_str);
    printf("⏱️  Time Taken: %.2f hours (excluding pauses)\n", hours);
    printf("⏸️  Total Pause Time: %.1f minutes\n", pause_time / 60.0);
    printf("📈 Average Speed: %s\n", speed_str);
    printf("============================================================\n");

    curl_global_cleanup();
    return 0;
}

File: net/net.py
#!/usr/bin/env python3
"""
High-speed data downloader converted from net.c to Python.

Features:
- Parallel downloads from multiple test URLs
- Keyboard controls: 'p' pause/resume, 'q' quit
- Progress bar with ETA, speed, total downloaded

This implementation uses `requests` and `threading`.
"""

from __future__ import annotations

import atexit
import random
import signal
import socket
import sys
import threading
import time
from typing import Optional
from urllib.parse import urlparse

try:
    from yt_dlp import YoutubeDL
except Exception:
    print("yt-dlp (python package 'yt-dlp') is required: pip install yt-dlp")
    raise
import os

try:
    import termios
    import tty
    import select
except Exception:
    termios = None  # type: ignore

import requests

# Disable SSL warnings for HTTPS requests in speed tests
import urllib3
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

# Default User-Agent to avoid 403s from simple servers
DEFAULT_UA = (
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/120.0.0.0 Safari/537.36"
)

# Mode: 'requests' (fast, streaming) or 'yt-dlp' (parsing-based)
MODE = "requests"


def is_resolvable(url: str) -> bool:
    """Return True if the URL's hostname can be resolved via DNS."""
    try:
        host = urlparse(url).hostname
        if not host:
            return False
        # Try resolving host (no port)
        socket.getaddrinfo(host, None)
        return True
    except Exception:
        return False


def filter_sources(sources: list[str]) -> list[str]:
    """Return only the sources whose hostnames resolve, printing warnings for others."""
    good = []
    for s in sources:
        if is_resolvable(s):
            good.append(s)
        else:
            print(f"⚠️  Skipping unreachable source (DNS failure): {s}")
    return good

# Target: 600GB
TARGET_GB = 600
TARGET_BYTES = TARGET_GB * 1024 * 1024 * 1024
CHUNK_SIZE = 512 * 1024  # 512KB (kept for compatibility but unused by yt-dlp path)
THREADS_PER_URL = 4

TEST_FILES = [
    "http://speedtest.tele2.net/10GB.zip",
    "http://speedtest.tele2.net/1GB.zip",
    "http://proof.ovh.net/files/10Gb.dat",
    "http://proof.ovh.net/files/1Gb.dat",
    "http://speedtest.belwue.net/100M",
    "http://speedtest-sgp1.digitalocean.com/10gb.test",
    "http://speedtest-ams2.digitalocean.com/10gb.test",
    "http://speedtest.ftp.otenet.gr/files/test10Gb.db",
]

NUM_URLS = len(TEST_FILES)
NUM_THREADS = NUM_URLS * THREADS_PER_URL

# Shared state
total_bytes = 0
total_lock = threading.Lock()
stop_event = threading.Event()
pause_event = threading.Event()
print_lock = threading.Lock()
source_failures = {}  # Track consecutive failures per source URL

# Terminal state
_old_termios = None


def format_bytes(num: int) -> str:
    units = ["B", "KB", "MB", "GB", "TB", "PB"]
    value = float(num)
    idx = 0
    while value >= 1024.0 and idx < len(units) - 1:
        value /= 1024.0
        idx += 1
    return f"{value:.2f} {units[idx]}"


def format_speed(bytes_per_sec: float) -> str:
    mbps = bytes_per_sec / (1024.0 * 1024.0)
    return f"{mbps:.1f} MB/s"


def format_eta(seconds: float) -> str:
    if seconds < 0 or seconds > 365 * 24 * 3600:
        return "--:--:--"
    hours = int(seconds / 3600)
    mins = int((seconds - hours * 3600) / 60)
    secs = int(seconds - hours * 3600 - mins * 60)
    return f"{hours:02d}:{mins:02d}:{secs:02d}"


def download_worker(url: str, thread_id: int) -> None:
    """Download a single URL using yt-dlp and report progress via progress hook.

    The hook receives incremental downloaded bytes and updates the global counter.
    """
    global total_bytes

    # per-thread last seen downloaded bytes (from yt-dlp progress dict)
    last_seen = {"bytes": 0}

    # rate-limit identical error messages per URL
    last_error_time = 0.0

    def progress_hook(d: dict) -> None:
        # Called frequently by yt-dlp with status updates
        # respect pause: block here while paused
        while pause_event.is_set() and not stop_event.is_set():
            time.sleep(0.1)

        if d.get("status") not in ("downloading",):
            return

        downloaded = d.get("downloaded_bytes") or 0

        # compute delta since last report for this thread
        delta = int(downloaded - last_seen["bytes"])
        if delta > 0:
            # update global total_bytes safely
            global total_bytes
            with total_lock:
                total_bytes += delta
            last_seen["bytes"] = downloaded

            # If target reached, signal stop to others and raise to abort
            if total_bytes >= TARGET_BYTES:
                stop_event.set()
                raise Exception("Target reached - aborting yt-dlp")

        # if caller requested stop, abort by raising
        if stop_event.is_set():
            raise Exception("Stopped by user")

    # Build options for yt-dlp. Write output to os.devnull to avoid disk usage.
    outtmpl = os.devnull
    ydl_opts = {
        "outtmpl": outtmpl,
        "noplaylist": True,
        "quiet": True,
        "no_warnings": True,
        "progress_hooks": [progress_hook],
        # retries help transient network errors
        "retries": 3,
    }

    attempt = 0
    while not stop_event.is_set():
        try:
            with YoutubeDL(ydl_opts) as ydl:
                # Reset attempt counter on success
                attempt = 0
                # yt-dlp raises on error; this call blocks until completion
                ydl.download([url])
        except Exception as exc:
            attempt += 1
            now = time.time()
            # Log at most once every 5 seconds per worker to avoid spam
            if now - last_error_time > 5:
                with print_lock:
                    print(f"ERROR (worker {thread_id}) for {url}: {exc}")
                last_error_time = now

            # Exponential backoff with jitter (cap 60s)
            backoff = min(60, (2 ** min(attempt, 6))) + random.random()
            time.sleep(backoff)
            continue


def requests_download_worker(url: str, part_idx: int, thread_id: int) -> None:
    """Download using HTTP streaming and discard data as it arrives (no disk usage).

    The worker attempts ranged requests if the server supports them, partitioning the
    file into THREADS_PER_URL parts. If ranges are unsupported, it streams the whole
    file and repeats until stopped/target reached.
    """
    session = requests.Session()

    last_error_time = 0.0
    attempt = 0
    # Ensure we reference the module-level counter
    global total_bytes

    while not stop_event.is_set():
        with total_lock:
            if source_failures.get(url, 0) >= 3:
                time.sleep(1)
                continue

        try:
            # Try HTTP first, fallback to HTTPS on 403
            current_url = url
            headers = {"User-Agent": DEFAULT_UA, "Referer": url}
            verify_ssl = not current_url.startswith("https://")  # Disable SSL verification for HTTPS to avoid cert issues in speed tests

            head = session.head(current_url, allow_redirects=True, timeout=10, headers=headers, verify=verify_ssl)
            if head.status_code == 403:
                # Fallback to HTTPS
                current_url = url.replace('http://', 'https://')
                headers["Referer"] = current_url
                verify_ssl = True  # For HTTPS, disable verification
                head = session.head(current_url, allow_redirects=True, timeout=10, headers=headers, verify=False)

            head.raise_for_status()
            cl = head.headers.get("content-length")
            length = int(cl) if cl and cl.isdigit() else None
            accept_ranges = head.headers.get("accept-ranges", "").lower() == "bytes"

            if length and accept_ranges:
                # Partition the file for parallel ranged downloads
                part_size = length // THREADS_PER_URL
                start = int(part_idx * part_size)
                end = int((start + part_size - 1) if part_idx < THREADS_PER_URL - 1 else length - 1)
                range_hdr = {"Range": f"bytes={start}-{end}", "User-Agent": DEFAULT_UA, "Referer": current_url}

                resp = session.get(current_url, headers=range_hdr, stream=True, timeout=15, verify=False if current_url.startswith("https://") else True)
                resp.raise_for_status()

                for chunk in resp.iter_content(chunk_size=CHUNK_SIZE):
                    if stop_event.is_set():
                        break
                    while pause_event.is_set() and not stop_event.is_set():
                        time.sleep(0.1)
                    if not chunk:
                        continue
                    with total_lock:
                        total_bytes += len(chunk)
                        if total_bytes >= TARGET_BYTES:
                            stop_event.set()
                            break
                # Finished this range; loop to start again (keeps bandwidth utilized)

            else:
                # Fallback: stream whole file (single-thread friendly)
                resp = session.get(current_url, headers=headers, stream=True, timeout=15, verify=False if current_url.startswith("https://") else True)
                resp.raise_for_status()
                for chunk in resp.iter_content(chunk_size=CHUNK_SIZE):
                    if stop_event.is_set():
                        break
                    while pause_event.is_set() and not stop_event.is_set():
                        time.sleep(0.1)
                    if not chunk:
                        continue
                    with total_lock:
                        total_bytes += len(chunk)
                        if total_bytes >= TARGET_BYTES:
                            stop_event.set()
                            break

            attempt = 0

        except Exception as exc:  # pragma: no cover - network behavior
            attempt += 1
            now = time.time()
            if now - last_error_time > 5:
                with print_lock:
                    print(f"ERROR (requests worker {thread_id}) for {url}: {exc}")
                last_error_time = now

            with total_lock:
                source_failures[url] = source_failures.get(url, 0) + 1

            backoff = min(60, (2 ** min(attempt, 6))) + random.random()
            time.sleep(backoff)
            continue


def keyboard_listener() -> None:
    # If termios isn't available, nothing to do
    if termios is None:
        return

    fd = sys.stdin.fileno()
    while not stop_event.is_set():
        dr, _, _ = select.select([sys.stdin], [], [], 0.1)
        if dr:
            c = sys.stdin.read(1)
            if not c:
                continue
            if c.lower() == "p":
                # toggle pause
                if pause_event.is_set():
                    pause_event.clear()
                    with print_lock:
                        print("\n▶️  RESUMED - Press 'p' to pause")
                else:
                    pause_event.set()
                    with print_lock:
                        print("\n⏸️  PAUSED - Press 'p' to resume")
            elif c.lower() == "q":
                with print_lock:
                    print("\n🛑 Stopping downloads...")
                stop_event.set()
                break


def restore_terminal() -> None:
    global _old_termios
    if _old_termios is not None and termios is not None:
        termios.tcsetattr(sys.stdin.fileno(), termios.TCSADRAIN, _old_termios)


def setup_terminal() -> bool:
    global _old_termios
    if termios is None:
        return False
    try:
        _old_termios = termios.tcgetattr(sys.stdin.fileno())
        new = _old_termios[:]
        # turn off canonical mode and echo
        new[3] = new[3] & ~(termios.ICANON | termios.ECHO)
        # VMIN = 0, VTIME = 0
        new[6][termios.VMIN] = 0
        new[6][termios.VTIME] = 0
        termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, new)
        atexit.register(restore_terminal)
        return True
    except Exception:
        return False


def print_progress(
    bytes_downloaded: int,
    total: int,
    speed: float,
    elapsed_hours: float,
    eta_seconds: float,
    paused: bool,
) -> None:
    bar_width = 30
    progress = float(bytes_downloaded) / float(total) if total > 0 else 0.0
    filled = int(progress * bar_width)

    bytes_str = format_bytes(bytes_downloaded)
    total_str = format_bytes(total)
    speed_str = format_speed(speed)
    eta_str = format_eta(eta_seconds)

    with print_lock:
        sys.stdout.write("\r")
        if paused:
            sys.stdout.write("⏸️  PAUSED : ")
        else:
            sys.stdout.write("📥 Progress: ")

        sys.stdout.write(f"{progress * 100:5.1f}%|")
        sys.stdout.write("".join("█" if i < filled else "░" for i in range(bar_width)))
        sys.stdout.write(
            f"| {bytes_str}/{total_str} | 🚀 {speed_str} | ⏱️ {elapsed_hours:.2f}h | ETA: {eta_str}"
        )
        if paused:
            sys.stdout.write(" (paused)")
        sys.stdout.write("    ")
        sys.stdout.flush()


def signal_handler(sig_num, frame) -> None:  # pragma: no cover - interactive
    stop_event.set()


def main() -> None:
    global total_bytes

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    term_ok = setup_terminal()
    if not term_ok:
        print("ℹ️  Keyboard controls not available in this terminal")

    print("============================================================")
    print("🚀 HIGH-SPEED DATA DOWNLOADER (Python Version)")
    print(f"🎯 Target: {TARGET_GB} GB")
    print(f"📡 Using up to {len(TEST_FILES)} configured sources (unreachable ones will be skipped) with {THREADS_PER_URL} threads each")
    print("============================================================")
    print("⚠️  Data is downloaded to RAM and discarded (no disk usage)")
    print("============================================================")
    print("🎮 Controls: Press 'p' to PAUSE/PLAY, 'q' to QUIT")
    print("============================================================\n")

    start_time = time.time()

    # Start keyboard listener thread
    kb_thread = threading.Thread(target=keyboard_listener, daemon=True)
    kb_thread.start()

    # Filter sources to reachable hosts and compute actual counts
    active_sources = filter_sources(TEST_FILES)
    if not active_sources:
        print("❌ No reachable sources found. Please check your network or modify TEST_FILES.")
        stop_event.set()
        return

    num_urls = len(active_sources)
    num_threads = num_urls * THREADS_PER_URL

    # Start download threads
    threads: list[threading.Thread] = []
    thread_id = 0
    for url in active_sources:
        for part_idx in range(THREADS_PER_URL):
            if MODE == "requests":
                t = threading.Thread(
                    target=requests_download_worker, args=(url, part_idx, thread_id), daemon=True
                )
            else:
                t = threading.Thread(
                    target=download_worker, args=(url, thread_id), daemon=True
                )
            t.start()
            threads.append(t)
            thread_id += 1

    print(f"Started {len(threads)} download threads using {num_urls} sources (mode={MODE})\n")

    last_bytes = 0
    avg_speed_smooth = 0.0
    pause_time = 0.0
    pause_start: Optional[float] = None

    while not stop_event.is_set():
        current_bytes = 0
        with total_lock:
            current_bytes = total_bytes

        # Check if target reached
        if current_bytes >= TARGET_BYTES:
            stop_event.set()
            break

        # Pause tracking
        if pause_event.is_set():
            if pause_start is None:
                pause_start = time.time()
        else:
            if pause_start is not None:
                pause_time += time.time() - pause_start
                pause_start = None

        elapsed = time.time() - start_time - pause_time
        elapsed_hours = elapsed / 3600.0 if elapsed > 0 else 0.0

        # speed = bytes in last second
        speed = float(current_bytes - last_bytes)

        if avg_speed_smooth == 0.0 and speed > 0:
            avg_speed_smooth = speed
        elif speed > 0:
            avg_speed_smooth = avg_speed_smooth * 0.9 + speed * 0.1

        eta_seconds = -1.0
        if avg_speed_smooth > 0:
            remaining = max(0, TARGET_BYTES - current_bytes)
            eta_seconds = remaining / avg_speed_smooth

        print_progress(
            current_bytes,
            TARGET_BYTES,
            speed,
            elapsed_hours,
            eta_seconds,
            pause_event.is_set(),
        )

        last_bytes = current_bytes
        time.sleep(1.0)

    # Wait for threads to finish
    print("\n\nWaiting for threads to finish...")
    for t in threads:
        t.join(timeout=0.1)

    kb_thread.join(timeout=0.1)

    final_bytes = 0
    with total_lock:
        final_bytes = total_bytes

    elapsed = time.time() - start_time - pause_time
    hours = elapsed / 3600.0 if elapsed > 0 else 0.0
    avg_speed = float(final_bytes) / elapsed if elapsed > 0 else 0.0

    print("\n============================================================")
    if final_bytes >= TARGET_BYTES:
        print(f"🎉 TARGET REACHED: {TARGET_GB} GB Downloaded!")
    else:
        print(f"📊 Downloaded: {format_bytes(final_bytes)}")
    print("============================================================")
    print(f"📥 Total Downloaded: {format_bytes(final_bytes)}")
    print(f"⏱️  Time Taken: {hours:.2f} hours (excluding pauses)")
    print(f"⏸️  Total Pause Time: {pause_time / 60.0:.1f} minutes")
    print(f"📈 Average Speed: {format_speed(avg_speed)}")
    print("============================================================")


if __name__ == "__main__":
    main()
